### DISCUSSION TITLE
How to fix "Chaining onto properties of binding state is deprecated" warnings?

### DISCUSSION THREAD

**Discussion by darrarski** [pointfreeco/swift-composable-architecture#2329] [Q&A] [answered]
I've recently updated ComposableArchitecture from `v0.54.1` to `v0.56.0` and noticed the following warnings appearing across my code:

> 'subscript(dynamicMember:)' is deprecated: Chaining onto properties of binding state is deprecated. Instead of pattern matching into a deeper property of binding state, use 'ReducerProtocol.onChange(of:)' to detect changes to nested properties of binding state. Instead of using 'viewStore.binding(\.$nested.property)', use dynamic member lookup ('viewStore.$nested.property').

At first sight, it's hard to guess which part of my code triggers it, because Xcode 14.3.1 (14E300c) provides just the file, without an exact line number.

After a brief investigation, I think the problem lays in the following code of my reducers:

```swift
Reduce { state, action in
  // ...
  case .binding(\.$query.dateRange): // <-- this line triggers the warning
    return .task { .loadInfo }
  // ...
}
```

and in my SwiftUI views:

```swift
struct DashboardFiltersView: View {
  let store: StoreOf<DashboardComponent>

  struct ViewState: Equatable {
    init(state: DashboardComponent.State) {
      query = state.query
      isLoading = state.isLoading
    }

    var query: DashboardInfoProvider.Query
    var isLoading: Bool
  }

  public var body: some View {
    WithViewStore(store, observe: ViewState.init) { viewStore in
      DayRangeFormView(
        range: viewStore.binding(
          get: \.query.dateRange,
          send: { DashboardComponent.Action.set(\.$query.dateRange, $0) } // <-- this line triggers the warning (probably)
        )
      )
      .disabled(viewStore.isLoading)
    }
  }
}
```

Despite the hints to use `ReducerProtocol.onChange(of:)` and dynamic member lookup, I'm not sure how the warnings can be fixed.

Will appreciate support from this amazing community :-)

**Comment by darrarski** [comment]
I think I found a way to silence the warnings that appeared in my reducers like this:

**Before:**

```swift
BindingReducer()

Reduce { state, action in
  // ...
  case .binding(\.$query.dateRange): // <-- this line triggers the warning
    return .task { .loadInfo }
  // ...
}
```

**After:**

```swift
BindingReducer()
  .onChange(of: \.query.dateRange) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }

Reduce { state, action in
  // ...
}
```

I'm not sure if this is the correct way though. It adds tons of boilerplate code, especially in places where I had multiple actions in pattern matching:

```swift
BindingReducer()

Reduce { state, action in
  // ...
  case .binding(\.$query.dateRange),
    .binding(\.$query.someProperty1),
    .binding(\.$query.someProperty2),
    .binding(\.$query.someProperty3):
    return .task { .loadInfo }
  // ...
}
```

becomes:

```swift
BindingReducer()
  .onChange(of: \.query.dateRange) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }
  .onChange(of: \.query.someProperty1) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }
  .onChange(of: \.query.someProperty2) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }
  .onChange(of: \.query.someProperty3) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }

Reduce { state, action in
  // ...
}
```

I still have no idea how to fix the warnings in my SwiftUI views.

**Reply by saroar** [reply]
Brandon Williams
 [3 days ago](https://pointfreecommunity.slack.com/archives/C04KQQ7NXHV/p1690200568438439?thread_ts=1690050715.976459&cid=C04KQQ7NXHV)
Also you cannot use .send(.set(\.$vCard.contact.firstName, …)). that should produce a deprecation warning. you must send the entire vCard in the action: .send(.set(\.$vCard, …)).

**Reply by mbrandonw** [reply, marked as answer]
We detailed all of these changes in [this discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions/2216), but the tldr is that while the previous syntax seemed great, there were a lot of gotchas that you can fall into, and people often did. So we had to change things.

Also, FWIW you _should_ be able to simplify your chain of `onChange`s with something like this:

```swift
BindingReducer()
  .onChange(
    of: { ($0.query.dateRange, $0.query.someProperty1, $0.query.someProperty2, $0.query.someProperty3) },
    removeDuplicates: ==
  ) { _, _ in
    Reduce { _, _ in 
      return .task { .loadInfo } 
    }
  }
```

However, it looks like we are missing the `removeDuplicates` variation of `onChange` and so it may be worth adding that. Too bad tuples aren't equatable in Swift :/

**Comment by saroar** [comment]
<img width="1352" alt="Screenshot 2023-07-27 at 15 26 03" src="https://github.com/pointfreeco/swift-composable-architecture/assets/8770772/35d3ced8-9d88-44a7-b076-a8f8fd27c685">

**Comment by JOyo246** [comment]
Our team is seeing the warning in many files across our app, despite only using binding chaining in 2 views. Some of the warnings are in files that don't even import TCA. Is this expected?

**Reply by mbrandonw** [reply]
Just sounds like a bug in Xcode. Often Xcode will put test failures, compile errors and deprecation warnings on the wrong line or on an entire file instead of the line. Have you tried fixing the warnings to see if that helps?

**Reply by salavert** [reply]
@JOyo246 I had the same problem and found this is due to batch mode, you can revert to the previous single file strategy with custom Build Setting `SWIFT_ENABLE_BATCH_MODE = NO` and it will pinpoint to the problematic file, or at least that's what solved my problem.
